android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerBinder.kt

```diff
@@ -39,9 +42,20 @@ object WorkerBinder {
    * @param worker The class that wants to be informed when to start and stop doing work.
    * @return [WorkerUnbinder] to unbind [Worker&#39;s][Worker] lifecycle.
    */
+  @OptIn(DelicateCoroutinesApi::class)
   @JvmStatic
-  open fun bind(interactor: Interactor<*, *>, worker: Worker): WorkerUnbinder {
-    return bind(mapInteractorLifecycleToWorker(interactor.lifecycle()), worker)
+  public fun bind(interactor: Interactor<*, *>, worker: Worker): WorkerUnbinder {
+    val workerEvents = interactor.lifecycleFlow.map { it.asWorkerEvent() }
+    // We need to launch in Unconfined dispatcher, AND start undispatched.
+    // If we only use Unconfined, and `onStart` launches some coroutine on Unconfined, coroutine will be put on an event loop.
+    val job = GlobalScope.launch(RibDispatchers.Unconfined, start = CoroutineStart.UNDISPATCHED) {
```

Comment:
This looks a little dangerous for me.  From my knowledge we expecting that `onStart` and `onStop` will be called on main/UI thread. I think it's OK to collect events on `Dispatchers.Unconfined` but we should switch to main in `onEach` and `onCompletion` blocks.

Reply:
In current `Worker` API, and still with this change, `onStart` and `onStop` will run on Main whenever `bind` is called on main. This commit does not change that. Note that calling `bind` on a worker thread already binds it (calls `onStart`) on a worker thread. There are 2 recent related PRs for defining the running thread for `onStart`. One introduces `RibCoroutineWorker` and other keeps `Worker` API but changes binding method.

Reply:
The `Context.Unconfined` doc it's doesn't guarantee that `onStart` and `onStop` will be called on the same thread as `bind` method was called. I also cannot understand how `CoroutineStart.UNDISPATCHED` will resolve looping. Referencing to the doc again: 

> However, when the coroutine is resumed from suspension it is dispatched according to the CoroutineDispatcher in its context

The resume methods in our case (`onEach`, `onCompletion`) will use `Context.Unconfined` and that start parameter will have affect only on `takeWhile` predicate, so it should be the same behavior with this `start =` parameter or without it . Maybe I misunderstood something :)

Reply:
Let's unwrap the flow (rough estimate):
```kotlin
flow {
  try { // onCompletion
    flow { // takeWhile 
      lifecycleFlow.collect {
        if (it == START) emit(it)
      }
    }.collect {
      emit(it)
    }
  } finally {
    worker.onStop()
  }
}.collect {
  worker.onStart()
}
```
Each inner emit immediately calls the lambda in the outer collect. Notice there is no suspension point whatsoever upon binding (except potentially `lifecycleFlow.collect` which I will discuss later), because `worker.onStart` and `worker.onStop` do not suspend. They will all be called synchronously in the context of the outermost collect.

Unconfined dispatcher docs state that it starts coroutine in the same call frame (similar to starting undispatched) and lets it resume from suspension on whatever thread called `continuation.resumeWith` after suspension. But we will never have any suspension when binding here, so this is irrelevant.

**How can it suspend?**

There's only one possibility in this code that makes it suspend upon binding: lifecycle flow has not emitted yet (lifecycle not started). It can be happen when trying to bind *before* `didBecomeActive`. In this corner case, we want `lifecycleFlow.collect` to resume immediately upon emission, and the **only** way to resume immediately, synchronously upon emission here is to do so on the thread that emitted the lifecycle event, which will be the main thread (see `Interactor`). For this resume immediately behavior, we need `Dispatches.Unconfined`.

**onStop**

I mentioned the flow will never suspend when binding (unless lifecycleFlow has no value), but what about unbinding? Obviously the coroutine will always suspend *after* binding, until the lifecycle terminates or user unbinds the worker.

How will resume happen in this case? If we use a dispatcher different than unconfined, the coroutine will not react immediately to the emission of the final lifecycle event. Instead the event delivery would be scheduled/dispatched. We rather have lifecycle collectors to react **immediately** to events.

**Why start both undispatched and in Unconfined**

I hope unconfined was justified by this point, but why also start undispatched?

Reply:
Great explanation, thanks for your time to write so detailed response 👍